rm/rm-server/src/main/java/org/ow2/proactive/resourcemanager/selection/topology/TopologyManager.java

```diff
@@ -585,19 +615,21 @@ public NodeSet select(int number, List<Node> matchedNodes) {
             NodeSet result = new NodeSet();
             for (Integer i : sortedCapacities) {
                 for (InetAddress host : hostsMap.get(i)) {
-                    List<Node> hostNodes = nodesOnHost.get(host);
-                    result.add(hostNodes.get(0));
-                    if (hostNodes.size() > 1) {
-                        List<Node> newExtraNodes = new LinkedList<>(hostNodes
-                                .subList(1, hostNodes.size()));
-                        if (result.getExtraNodes() == null) {
-                            result.setExtraNodes(new LinkedList<Node>());
+                    LinkedHashSet<Node> hostNodes = nodesOnHost.get(host);
+                    if (hostNodes.size() > 0) {
+                        result.add(hostNodes.iterator().next());
+                        if (hostNodes.size() > 1) {
+                            List<Node> newExtraNodes = new LinkedList<>(subListLHS(hostNodes,
+                                    1, hostNodes.size()));
+                            if (result.getExtraNodes() == null) {
+                                result.setExtraNodes(new LinkedList<Node>());
+                            }
+                            result.getExtraNodes().addAll(newExtraNodes);
+                        }
+                        if (--number <= 0) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)